### PR TITLE
[fix][broker] Fix the reason label of authentication metrics

### PR DESCRIPTION
**Pull request by gaoran10**
### Motivation

The `reason` label of authentication metrics is a potential infinity string.
The PR https://github.com/apache/pulsar/pull/19758 also tries to fix this problem.

Use an enum ErrorCode as the authentication failure reason.

### Modifications

Add an enum class `ErrorCode` for Pulsar authentication, it indicates the authentication failure reason.
Add a new exception `PulsarAuthenticationException` which contains the error code.

### Verifying this change

Change the test to match this change.

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/gaoran10/pulsar/pull/26

### CODE REVIEW

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderBasic.java
Comment (mattisonchao) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderBasic.java:112]: IMO, use `ErrorCode.BASIC_INVALID_AUTH_DATA` directly will be more clear. :)

**Review by mattisonchao — APPROVED** [review]

**Comment by mattisonchao** [comment]
@michaeljmarshall would you mind to taking a look here?

**Review by codelipenghui — APPROVED** [review]
The change looks good me.
And I have added the release labels. Since it's a metrics changes that might impact the Pulsar operators, maybe not, I don't know someone will add alert rules or something based on the long reason message. But it still worth to start a discussion for cherry-picking the changes to release branches.

**Comment by codelipenghui** [comment]
And please also update the documentation after the RP get merged.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java
Comment (michaeljmarshall) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java:161]: I think it would be valuable to increment the metric outside of the implementations of the `AuthenticationProvider` so that all custom implementations can integrate with this feature.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/PulsarAuthenticationException.java
Comment (michaeljmarshall) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/PulsarAuthenticationException.java:46]: My primary concern with relying on this enum is that it is not easily extended. For example, I just merged the OIDC Provider with the following enum:

https://github.com/apache/pulsar/blob/bafecb2a9c0e73942de6a38df72dd5888d5afd66/pulsar-broker-auth-oidc/src/main/java/org/apache/pulsar/broker/authentication/oidc/AuthenticationExceptionCode.java#L25-L38

From your list and from mine, my primary conclusion is that the types of errors are often provider specific, so a static enum that cannot be extended feels restrictive.

I agree that it is extremely important to limit the cardinality of these metrics, but I also think we want to make it easily extensible by users.

I think an alternative solution could work by making `PulsarAuthenticationException` abstract and letting each provider implement their own exception. Then, the provider itself could have its own enum, and when the framework calls `getErrorCode`, the implementation will convert its own enum into a string.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/metrics/AuthenticationMetrics.java
Comment (michaeljmarshall) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/metrics/AuthenticationMetrics.java:64]: Because this is a generic metrics class meant to be useful to all authentication provider implementations, I think we should provide an easy way to provide custom "reasons". I respect that it is easy to provide a reason that is high cardinality, but I think we should expect each implementation to handle those details.

**Review by michaeljmarshall — CHANGES_REQUESTED** [review]
Thanks for working on this @gaoran10, I just noticed this problem last week. It's definitely one we need to fix.

My primary concern is with the `PulsarAuthenticationException` class because it isn't really extensible for custom implementations of the `AuthenticationProvider` interface.

I also wonder if it would be better to add a new class that makes it so we can avoid each implementation from needing to increment the metrics on its own. As long as a provider throws a `PulsarAuthenticationException`, we could increment a metric in the broker/proxy/function worker/websocket proxy.

Let me know what you think. Thanks!

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/PulsarAuthenticationException.java
Comment (gaoran10) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/PulsarAuthenticationException.java:46]: Good idea! I'll change this.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/metrics/AuthenticationMetrics.java
Comment (gaoran10) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/metrics/AuthenticationMetrics.java:64]: I add an abstract class `BaseAuthenticationException`, which extends `AuthenticationException` and has an abstract method `public abstract Enum<?> getErrorCode()`, I think developers can custom their exceptions based on the `BaseAuthenticationException`, the custom exception can provide the custom error code.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java
Comment (gaoran10) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java:162]: It seems it's hard to find a unified entrance to increment the authentication metric. Maybe we can try to improve this in another PR. WDYT

**Comment by gaoran10** [comment]
@michaeljmarshall PTAL

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/PulsarAuthenticationException.java
Comment (michaeljmarshall) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/PulsarAuthenticationException.java]: My primary concern is that `ErrorCode` has many enums that are specific to each provider and are not generally useful. What do you think about the following class?

```java
    public class PulsarAuthenticationException extends AuthenticationException {

        private final String errorCode;

        // Warning, the errorLabel must not have high cardinality
        public PulsarAuthenticationException(String message, String errorLabel) {
            super(message);
            this.errorLabel = errorLabel;
        }

        public String getErrorLabel() {
            return errorLabel;
        }

    }
```

I think it is reasonable to expect maintainers to know that the `errorLabel` must not be high cardinality, and this way, we don't have enumerations that combine unrelated labels. If a provider wants to use an enum for its own tracking, that would be perfectly reasonable, and it would probably work just like the `AuthenticationProviderOpenID`.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java
Comment (michaeljmarshall) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java:170]: Is it possible to increment the `INVALID_AUTH_DATA` error in the `getToken` method?

**Review by michaeljmarshall — DISMISSED** [review]
Thanks for making the changes @gaoran10. I left a couple comments.

File: pulsar-broker-auth-oidc/src/main/java/org/apache/pulsar/broker/authentication/oidc/AuthenticationProviderOpenID.java
Comment (gaoran10) [reply pulsar-broker-auth-oidc/src/main/java/org/apache/pulsar/broker/authentication/oidc/AuthenticationProviderOpenID.java:92]: Ok, I'll revert this change.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java
Comment (gaoran10) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java:170]: The method `getToken` is static, some methods use it and catch exceptions to record failure metrics, such as OpenID provider, if we increment the `INVALID_AUTH_DATA` error in the `getToken` method, this will generate repeat failure records.

OpenID provider
```
    CompletableFuture<DecodedJWT> authenticateTokenAsync(AuthenticationDataSource authData) {
        String token;
        try {
            token = AuthenticationProviderToken.getToken(authData);
        } catch (AuthenticationException e) {
            incrementFailureMetric(AuthenticationExceptionCode.ERROR_DECODING_JWT);
            return CompletableFuture.failedFuture(e);
        }
        return authenticateToken(token)
                .whenComplete((jwt, e) -> {
                    if (jwt != null) {
                        AuthenticationMetrics.authenticateSuccess(getClass().getSimpleName(), getAuthMethodName());
                    }
                    // Failure metrics are incremented within methods above
                });
    }

```

File: pulsar-broker-auth-oidc/src/main/java/org/apache/pulsar/broker/authentication/oidc/AuthenticationProviderOpenID.java
Comment (gaoran10) [reply pulsar-broker-auth-oidc/src/main/java/org/apache/pulsar/broker/authentication/oidc/AuthenticationProviderOpenID.java]: Ok, I changed to use the `AUTH_METHOD_NAME`(token). Thanks.

**Review by michaeljmarshall — COMMENTED** [review]
Thanks for addressing my feedback @gaoran10. Would you mind getting the build to pass? Then I'll re-review it. Thanks!

**Comment by lhotari** [comment]
> Thanks for addressing my feedback @gaoran10. Would you mind getting the build to pass? Then I'll re-review it. Thanks!

@michaeljmarshall when changes are requested, the build in upstream won't run at all. I'll dismiss the "changes requested" to get the tests running.

**Comment by lhotari** [comment]
> > Thanks for addressing my feedback @gaoran10. Would you mind getting the build to pass? Then I'll re-review it. Thanks!
> 
> @michaeljmarshall when changes are requested, the build in upstream won't run at all. I'll dismiss the "changes requested" to get the tests running.

I should have just added the "ready-to-test" label to the PR instead of dismissing the review. I forgot that it's the option that I should have used.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java
Comment (michaeljmarshall) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java:170]: Thank for your explanation.

**Review by michaeljmarshall — APPROVED** [review]
LGTM

**Comment by michaeljmarshall** [comment]
As discussed on the mailing list https://lists.apache.org/thread/w4jzk27qhtosgsz7l9bmhf1t7o9mxjhp, there is no plan to release 2.9.6, so I am going to remove the release/2.9.6 label

**Pull request merged**
